test/test_torch.py

```diff
@@ -5995,6 +5993,54 @@ def test_grad_scale_will_not_overflow(self, device):
         scaler.update()
         assert scaler._scale != float("inf") and scaler._scale != float("nan")
 
+    @onlyNativeDeviceTypes
+    @skipIfTorchDynamo("https://github.com/pytorch/pytorch/issues/123238")
+    @dtypes(*floating_types_and(torch.bfloat16, torch.half))
+    @precisionOverride({torch.half : 1e-3, torch.bfloat16 : 1e-3})
+    def test_fused_adam_and_adamw(self, device, dtype):
+        r"""
+        This testcase will compare the results between
+        1. _single_tensor_adam and _fused_adam.
```

Comment:
This is what test_fused_matches_forloop does in test/test_optim.py

Reply:
I wish to use this UT to just compare `_single_tensor_adam/adamw` and ` _fused_adam/adamw` while `test_fused_matches_forloop` have a larger scope ( with more functionality with `fw/bw, etc`). And I saw `test_fused_matches_forloop` only test `float64`.

However, the tolerance in this `test_fused_adam_and_adamw` may also too high for `bfloat16/float16`, the "diff" comes from the float/bfloat16` cast. The non-fused kernel have more cast than fused kernel.
For example
```
// a, b, c are bfloat16 tensor
// to perform a+b*c
# non-fused-kernel
(a.bfloat().float() + (b.float() * c.float()).bfloat16().float()).bfloat16()
# fused-kernel
(a.float() + b.float() * c.float()).bfloat16()
```
I should enhance `test_fused_adam_and_adamw` by cast all tensor args to `float` and run `_single_tensor_adam`, then cast all tensors args back to `bfloat16`, and then compare with `fused_adam`.
But with `test_fused_matches_forloop`, it is harder to do this and the tolerance will be very large to pass the CI since it run 7 iters. The `diff` can be accumulate. Do you have suggestions on this?

Reply:
Ah, for the 7 iterations diff building up--I built TensorTracker https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_optimizers.py#L2220 which will do the work of checking for equality + then aligning the values so that the discrepancies don't build up too much. Example usage is here: https://github.com/pytorch/pytorch/blob/main/test/test_optim.py#L497.

This may allow expanding the one test to test all the functionalities you wanna test.

Reply:
Thanks, I removed `test_fused_adam_and_adamw` and use `test_fused_matches_forloop` for `float64, float32, bfloat16, float16`. And also use then `TensorTracker` you mentioned.